article.md

```diff
@@ -0,0 +1,612 @@
+|                 | createdAt: date                                | updatedAt: date  |                   |
+|                 | updatedAt: date                                |                  |                   |
+
+Let's break down the associations. A `User` has many `Post`'s which they have authored. This is the `AuthorId` column on `Posts`. This is a one-to-many relationship. A `User` also has a many-to-many relationship with `Posts` via `PostUpvotes`: a user can upvote many posts, and a post can be upvoted by many users. Finally, `Comments` have both a `User`, via `AuthorId`, and a `Post`. These are both one-to-many.
+
+## Finally, some code
+
+Let's create a simple ExpressJS server. This article will follow along with [this][repo] repo. We use tags so you can see what the code should look
+like at each stage. Let's clone the initial stage, which is just a very simple Express server written in Typescript.
+
+```shell
+$ git clone git@github.com:ahmerb/ts-sqlize-code.git
+$ cd ts-sqlize-code
+$ git checkout 0.start
+```
+
+We can run the app with
+
+```shell
+$ yarn install
+$ yarn start
+```
+
+We now have a very simple Express app. Take a look at `app.ts`. It just defines a single request handler that returns the json `{ message: 'hello, world' }` and then tells Express to listen on port 3000.
+
+## Models, models, models
+
+Now, we want to setup Sequelize models that reflect the db schema we defined earlier. We will create a file for each model in the directory `src/models/`. Each file will export two types: `ModelAttributes` and `ModelInstance`. They'll also export a `ModelFactory` function which will actually set up the model and register it with Sequelize. At first, we're going to ignore associations. We'll deal with these later.
+
+Let's start with the `Users` model. Firstly, we'll define an interface `UserAttributes`. This interface defines all the attributes you specify when creating a new instance of the model. In our database schema, we defined that a `User` has `id`, `name`, `createdAt` and `updatedAt` attributes. Let's see how that looks.
+
+```typescript
+// src/models/User.ts
+export interface UserAttributes {
+  id?: number;
+  name: string;
+  createdAt?: Date;
+  updatedAt?: Date;
+};
+```
+
+Simple. Note that `id`, `createdAt` and `updatedAt` are optional. This is because, when creating a new instance of a model, we don't want to be
+forced to specify these fields. Instead, we usually want Sequelize to automatically set them.
+
+The second type each model needs is a `ModelInstance`. This represents a Sequelize instance for an actual database row. For example, the function `User.find({ id: 1 })` returns an instance of the `User` model. It will have the `ModelAttributes` we defined (i.e., the table columns) as well as more Sequelize instance methods such as `.validate`, `.save`, `.update`, et cetera. For `Users`, the interface is defined as follows.
```

Comment:
Maybe briefly mention somewhere that for complex querying, raw SQL snippets might be required?

Reply:
I think that's a statement that applies for using any ORM, whether that's Sequelize, ActiveRecord (rails/ruby), Entity (asp.net/c#), etc